test/jdk/java/net/httpclient/FileChannelPublisherTest.java

```diff
@@ -0,0 +1,693 @@
+                    pair, tempDir, fileLength, fileChannelOffset, fileChannelLength);
+        }
+
+        {
+            int fileChannelOffset = 2;
+            int fileChannelLength = 2 * DEFAULT_BUFFER_SIZE;
+            String debuggingContext = debuggingContext(fileLength, fileChannelOffset, fileChannelLength);
+            assertTrue(
+                    fileLength - fileChannelOffset > fileChannelLength,
+                    "must end before EOF " + debuggingContext);
+            testSuccessfulContentDelivery(
+                    "Partial content *before* the EOF. Occupies exactly 2 buffers. " + debuggingContext,
+                    pair, tempDir, fileLength, fileChannelOffset, fileChannelLength);
+        }
+
+        {
+            int fileChannelOffset = 2;
+            int fileChannelLength = 3 * DEFAULT_BUFFER_SIZE - 2;
+            String debuggingContext = debuggingContext(fileLength, fileChannelOffset, fileChannelLength);
+            assertTrue(
+                    fileLength - fileChannelOffset > fileChannelLength,
+                    "must end before EOF " + debuggingContext);
+            testSuccessfulContentDelivery(
+                    "Partial content *before* the EOF. Occupies 3 buffers, the last is custom sized. "+ debuggingContext,
+                    pair, tempDir, fileLength, fileChannelOffset, fileChannelLength);
+        }
+
+    }
+
+    private static String debuggingContext(int fileLength, int fileChannelOffset, int fileChannelLength) {
+        Map<String, Object> context = new LinkedHashMap<>();    // Using `LHM` to preserve the insertion order
+        context.put("DEFAULT_BUFFER_SIZE", DEFAULT_BUFFER_SIZE);
+        context.put("fileLength", fileLength);
+        context.put("fileChannelOffset", fileChannelOffset);
+        context.put("fileChannelLength", fileChannelLength);
+        boolean customSizedBuffer = fileChannelLength % DEFAULT_BUFFER_SIZE == 0;
+        context.put("customSizedBuffer", customSizedBuffer);
+        return context.toString();
+    }
+
+    private void testSuccessfulContentDelivery(
+            String caseDescription,
+            ServerRequestPair pair,
+            Path tempDir,
+            int fileLength,
+            int fileChannelOffset,
+            int fileChannelLength) throws Exception {
+
+        // Case names come handy even when no debug logging is enabled.
+        // Hence, intentionally avoiding `Logger`.
+        System.err.printf("Case: %s%n", caseDescription);
+
+        // Create the file to upload
+        String fileName = "data-%d-%d-%d.txt".formatted(fileLength, fileChannelOffset, fileChannelLength);
+        Path filePath = tempDir.resolve(fileName);
+        withFileChannel(filePath, fileLength, (fileBytes, fileChannel) -> {
+
+            // Upload the file
+            HttpRequest request = pair
+                    .requestBuilder
+                    .POST(ofFileChannel(fileChannel, fileChannelOffset, fileChannelLength))
```

Comment:
I think it would be better to not add a static import for this method and instead use `BodyPublishers.ofFileChannel(...)` to make it easily visible what this method is (especially because this test has a similarly named `withFileChannel(...)` method).

Reply:
Changed in 99503c0bac7.